### DISCUSSION TITLE
Connection pooling with Postgres through Supabase

### DISCUSSION THREAD

**Discussion by LGmatrix13** [vercel/community#1283] [Help] [answered]
### Summary

Using nextjs api routes with the node-postgres package. Having some connection issues with my database. I'm also using connection pooling on supabase. Here's what my connection looks like:

```
import { Pool } from "pg";

let client;

if (!client) {
  client = new Pool({
    host: process.env.PGHOST,
    user: process.env.PGUSER,
    port: process.env.PGPORT,
    password: process.env.PGPASSWORD,
    database: process.env.PGDATABASE,
  });
}

export default client;

```

Connection pooling should help with the issue, but it does not look that way. I'm the only one using the app, and it fails to connect 50% of the time. Running production locally does not have any issues, probably because it is a single-node instance.

Any suggestions? Thank you!

**Comment by leerob** [comment]
Hey! If you're using Supabase, you should use the JavaScript client library, which will take care of managing connections for you.

- https://supabase.com/docs/guides/getting-started/quickstarts/nextjs
- https://vercel.com/guides/using-databases-with-vercel

**Reply by LGmatrix13** [reply]
I would, but I'm only using supabase to get up and running. Plan to self-host an instance on digital ocean later...

When trying a connection with a neon.tech database, the same thing occurs...

**Reply by jtlapp** [reply]
@leerob, Can you elaborate on why we "should" use the Supabase client library? It doesn't support transactions, doesn't work with any ORMs, and doesn't work with schema-safe query builders like Kysely.

**Reply by bombillazo** [reply]
As mentioned, the supabase client is an abstraction for PostgREST, so its bound to those features, hence the lack of transactions or raw query running. Id love to use it in our back end for more custom/advanced stuff but its simply not possible _only_ with Supabase...

**Comment by LGmatrix13** [comment, marked as answer]
Had to turn on connection pooling in supabase and then replace my connection string: https://supabase.com/docs/guides/database/connecting-to-postgres#connection-pool

**Reply by homocodian** [reply]
Which connection pooling method from supabase (Session || Transaction) should I use for NextJs 14 API Routes with Drizzle ORM?

**Reply by rajeshdavidbabu** [reply]
According to the official docs "Transaction Mode", meaning your DB string should point to the port 6543

https://orm.drizzle.team/docs/get-started-postgresql#supabase

**Comment by adampotency** [comment]
In my case I entered the env vars manually based on the database url rather than connecting the project via vercel (which had all the correct env vars that made use of connection pooling etc)
